### PR TITLE
Geth 1.8.1 integration test & compatibility

### LINKED ISSUE #655: getTransactionReceipt propagates error from Geth v1.8.0 when there's no receipt yet
* Version: 4.0.0b7
* Python: 3.5.2
* OS: Ubuntu 16.04

### What was wrong?

I am calling web3.eth.getTransactionReceipt of a hash that doesn't have a receipt yet and as a result it propagates the error from Geth instead of returning None.
I am connected to a private node with Geth v1.8.0.

This is what I get:

```
>>> web3.eth.getTransactionReceipt(txHash)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.5/dist-packages/web3/eth.py", line 188, in getTransactionReceipt
    [transaction_hash],
  File "/usr/local/lib/python3.5/dist-packages/web3/manager.py", line 102, in request_blocking
    raise ValueError(response["error"])
ValueError: {'code': -32000, 'message': 'unknown transaction'}
```

### PR DESCRIPTION
**Pull request by carver**
### What was wrong?

Fixes #655 - `getTransactionReceipt` raising an exception

### How was it fixed?

Catch `ValueError` and return None, on `eth_getTransactionReceipt`

TODO:

- Make new issue for more fine-grained error code capture and normalization, by node type
- document new default middleware

#### Cute Animal Picture

Cute animal picture

### CODE REVIEW

**Comment by carver** [comment]
`generate_go_ethereum_fixture.py` is really sweet @pipermerriam -- super straightforward

Currently broken because py-geth doesn't support 1.8.1 yet. Once https://github.com/ethereum/py-geth/pull/42 passes, I'll merge and deploy `py-geth`, and then re-run these tests.

**Comment by carver** [comment]
So, this is weird... (from `geth attach` with the geth fixtures)

```js
> web3.eth.getBlock('latest')
{
  difficulty: 1,
  extraData: "0x7765623370792d746573742d636861696e",
  gasLimit: 4707788,
  gasUsed: 0,
  hash: "0xf312697fa82fcf4a36e956aec17bf52eaaf15f6a4a30ac8cb7417aec1d076cf9",
  logsBloom: "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  miner: "0x3333333333333333333333333333333333333333",
  mixHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
  nonce: "0xdeadbeefdeadbeef",
  number: 0,
  parentHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
  receiptsRoot: "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  sha3Uncles: "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  size: 521,
  stateRoot: "0x8d1019bdf25a139a57ff2c4e2f92abec83a45bf5b00cca5ca8c7c48a12b6704d",
  timestamp: 0,
  totalDifficulty: 1,
  transactions: [],
  transactionsRoot: "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  uncles: []
}
> web3.eth.getBlock(1)
{
  difficulty: 131072,
  extraData: "0xd783010801846765746887676f312e392e34856c696e7578",
  gasLimit: 4712384,
  gasUsed: 200511,
  hash: "0xd329fd7a79c145171fd7965bc107eff705bd93554f33132aa54342c83335be96",
  logsBloom: "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  miner: "0xdc544d1aa88ff8bbd2f2aec754b1f1e99e1812fd",
  mixHash: "0xa88637b953be3f00070bb5a4afb9b70b3ebf0213c77a7432518cc4a27aff8b59",
  nonce: "0x7bf65f8747d192b7",
  number: 1,
  parentHash: "0xf312697fa82fcf4a36e956aec17bf52eaaf15f6a4a30ac8cb7417aec1d076cf9",
  receiptsRoot: "0xcc4a82d9c64330182623bb1fb9e671184c7064a26e546ff4198d944eb7ae9aa9",
  sha3Uncles: "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  size: 1208,
  stateRoot: "0x0ddb48366c4ab8893fb9d112b2934b15c77af8f13a36c8eb6959d25a5c647d32",
  timestamp: 1519429173,
  totalDifficulty: 131073,
  transactions: ["0x20775bfaf519f8f0c932c1bd256fc954de707ef5c6c102564d01e2605409a48c"],
  transactionsRoot: "0x4887f75801f02147293caee913b17b4988c83f2c1ff5dee1f2536da7206e0d51",
  uncles: []
}
> web3.eth.getTransactionReceipt('0x20775bfaf519f8f0c932c1bd256fc954de707ef5c6c102564d01e2605409a48c')
{
  blockHash: "0xd329fd7a79c145171fd7965bc107eff705bd93554f33132aa54342c83335be96",
  blockNumber: 1,
  contractAddress: "0xd794c821fccff5d96f5db44af7e29977630a9dc2",
  cumulativeGasUsed: 200511,
  from: "0xdc544d1aa88ff8bbd2f2aec754b1f1e99e1812fd",
  gasUsed: 200511,
  logs: [],
  logsBloom: "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  root: "0x734f97176bb5812e71b1099331651ba51189c7b19d6901bc0365e6893a91d691",
  to: null,
  transactionHash: "0x20775bfaf519f8f0c932c1bd256fc954de707ef5c6c102564d01e2605409a48c",
  transactionIndex: 0
}
```

Why does the `latest` block have lower difficulty than block 1? Why does the transaction receipt return as included in the block, even though the block doesn't appear to be on the canonical chain?

**Comment by carver** [comment]
Solved. We needed --gcmode=archive to guarantee that all blocks will be written to disk before shutdown. See equivalent problem (but during import) at https://github.com/ethereum/go-ethereum/issues/16097

At first I thought there was some race condition with `geth` shutting down after copying the block files out, so that part of the fixture generation is cleaned up, too.

There is also a new verification step to make sure that the latest block is available after shutting down and reopening geth. The generation script will raise an assertion error if this fails.

File: web3/middleware/normalize_errors.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+from cytoolz import (
+    assoc,
+    dissoc,
+)
+
+
+def normalize_errors_middleware(make_request, web3):
+    def middleware(method, params):
+        result = make_request(method, params)
+        if method == 'eth_getTransactionReceipt' and 'error' in result:
```
Comment (pipermerriam) [inline web3/middleware/normalize_errors.py]: This block probably deserves a comment.  Code reads pretty well, but a nice comment explaining why will be unambiguous.

**Comment by carver** [comment]
> I assume the error normalization middleware arose from seeing the integration tests fail.

Unfortunately, no. Our first report was #655 because we hadn't built the geth 1.8 integration tests yet.

But, the integration test *did* catch a bug in my first implementation. So the integration tests still get a +1 from me.

**Comment by carver** [comment]
Ah, that's one thing that's been on my to-do list: get circle to test against the merged PR instead of the PR tip. That's why circle had a false positive, I believe.

**Pull request merged**